File: DESYJun24/convert_feather_to_root.cpp
#include <iostream>
#include <vector>
#include <map>
#include <ROOT/RDataFrame.hxx>
#include <TFile.h>
#include <TTree.h>
#include <arrow/api.h>
#include <arrow/io/api.h>
#include <arrow/ipc/api.h>

arrow::Status ConvertFeatherToRoot(const std::string& featherFile, const std::string& rootFile) {
    // Open Feather file
    std::shared_ptr<arrow::io::ReadableFile> infile;
    ARROW_ASSIGN_OR_RAISE(infile, arrow::io::ReadableFile::Open(featherFile));

    std::shared_ptr<arrow::ipc::RecordBatchFileReader> reader;
    ARROW_ASSIGN_OR_RAISE(reader, arrow::ipc::RecordBatchFileReader::Open(infile));

    std::vector<std::shared_ptr<arrow::RecordBatch>> batches;
    for (int i = 0; i < reader->num_record_batches(); i++) {
        ARROW_ASSIGN_OR_RAISE(auto batch, reader->ReadRecordBatch(i));
        batches.push_back(batch);
    }

    // Combine batches into a table
    std::shared_ptr<arrow::Table> table;
    ARROW_ASSIGN_OR_RAISE(table, arrow::Table::FromRecordBatches(batches));

    // Extract column indices
    std::shared_ptr<arrow::ChunkedArray> evt_column = table->GetColumnByName("evt");
    if (!evt_column) {
        std::cerr << "Column 'evt' not found!" << std::endl;
        return arrow::Status::OK();
    }

    // Open ROOT file and create TTree
    TFile f(rootFile.c_str(), "RECREATE");
    TTree tree("Events", "Tree storing feather data");

    std::map<int, std::vector<int>> evt_data;
    std::map<std::string, std::map<int, std::vector<int>>> int_columns;
    std::map<std::string, std::map<int, std::vector<float>>> float_columns;

    // Column names
    std::vector<std::string> int_col_names = {"bcid", "l1a_counter", "ea", "board", "row", "col", "toa", "tot", "cal"};
    std::vector<std::string> float_col_names = {"acal"};

    // Read columns into maps
    for (const auto& name : int_col_names) {
        int_columns[name] = {};
        auto col = table->GetColumnByName(name);
        if (!col) continue;
        auto int_array = std::static_pointer_cast<arrow::Int64Array>(col->chunk(0));
        for (int i = 0; i < int_array->length(); i++) {
            int evt = evt_column->chunk(0)->GetScalar(i).ValueOrDie()->hash();
            int val = int_array->Value(i);
            int_columns[name][evt].push_back(val);
        }
    }

    for (const auto& name : float_col_names) {
        float_columns[name] = {};
        auto col = table->GetColumnByName(name);
        if (!col) continue;
        auto float_array = std::static_pointer_cast<arrow::DoubleArray>(col->chunk(0));
        for (int i = 0; i < float_array->length(); i++) {
            int evt = evt_column->chunk(0)->GetScalar(i).ValueOrDie()->hash();
            float val = static_cast<float>(float_array->Value(i));
            float_columns[name][evt].push_back(val);
        }
    }

    // Create branches in the tree
    std::vector<int> evt_vector;
    std::map<std::string, std::vector<int>> int_vectors;
    std::map<std::string, std::vector<float>> float_vectors;

    tree.Branch("evt", &evt_vector);
    for (const auto& name : int_col_names) {
        tree.Branch(name.c_str(), &int_vectors[name]);
    }
    for (const auto& name : float_col_names) {
        tree.Branch(name.c_str(), &float_vectors[name]);
    }

    // Fill the tree
    for (const auto& [evt, _] : evt_data) {
        evt_vector = {evt};
        for (const auto& name : int_col_names) {
            int_vectors[name] = int_columns[name][evt];
        }
        for (const auto& name : float_col_names) {
            float_vectors[name] = float_columns[name][evt];
        }
        tree.Fill();
    }

    tree.Write();
    f.Close();
    std::cout << "ROOT file created successfully: " << rootFile << std::endl;

    return arrow::Status::OK();
}

int main(int argc, char *argv[]) {
    if (argc != 3) {
	std::cout << "You must enter two arguments" << std::endl; 
	return 1;
    }
    std::string inFile = argv[1];
    std::string rootFile = argv[2];
    arrow::Status status = ConvertFeatherToRoot(inFile, rootFile);
    if (!status.ok()) {
        std::cerr << "Error: " << status.message() << std::endl;
        return 1;
    }
    return 0;
}

File: DESYJun24/convert_feather_to_root.py
import ROOT as R
import numpy as np
import pandas as pd
import time
from tqdm import tqdm
import os
from argparse import ArgumentParser

parser = ArgumentParser()
parser.add_argument("--dataset", default="0")

args = parser.parse_args()
DATA  = f"datos/loop_{args.dataset}.feather"

''' 
Require events to have exactly one hit in each board
And to have CALcode values that are close to the average CALcode
'''
def filterEvents(event):
    if not len(event) == 4: return False
    if not len(event['board']) == len(np.unique(event['board'])): return False
    return True

# Function to write Feather file data to ROOT
def feather_to_root(feather_file, root_file):
    # Read the Feather file into a pandas DataFrame
    df = pd.read_feather(feather_file)

    # Open the ROOT file for writing
    root_file = R.TFile(root_file, "RECREATE")

    # Create a TTree object
    tree = R.TTree("data_tree", "Tree storing event data")

    # Get the unique events (evt)
    events = df['evt'].unique()
    print(events)

    # Create ROOT branches for each variable
    bcid = R.std.vector('int')()
    l1a_counter = R.std.vector('int')()
    ea = R.std.vector('int')()
    board = R.std.vector('int')()
    row = R.std.vector('int')()
    col = R.std.vector('int')()
    toa = R.std.vector('int')()
    tot = R.std.vector('int')()
    cal = R.std.vector('int')()

    # Add branches to the tree
    #tree.Branch('evt', R.std.vector('int')()) # evt will be a single int for each entry
    tree.Branch('bcid', bcid)
    tree.Branch('l1a_counter', l1a_counter)
    tree.Branch('ea', ea)
    tree.Branch('board', board)
    tree.Branch('row', row)
    tree.Branch('col', col)
    tree.Branch('toa', toa)
    tree.Branch('tot', tot)
    tree.Branch('cal', cal)

    # Iterate over each unique event 'evt'
    for evt in tqdm(events, total=max(events)):
        # Filter the dataframe for this event
        event_data = df[df['evt'] == evt]
        if not len(event_data) == 4: continue
        if not len(event_data['board']) == len(np.unique(event_data['board'])): continue
        #print(event_data)
        
        # Clear the vectors for the current event
        bcid.clear()
        l1a_counter.clear()
        ea.clear()
        board.clear()
        row.clear()
        col.clear()
        toa.clear()
        tot.clear()
        cal.clear()

        # Fill the vectors with the values for this event
        for index, row_data in event_data.iterrows():
            bcid.push_back(int(row_data['bcid']))
            l1a_counter.push_back(int(row_data['l1a_counter']))
            ea.push_back(int(row_data['ea']))
            board.push_back(int(row_data['board']))
            row.push_back(int(row_data['row']))
            col.push_back(int(row_data['col']))
            toa.push_back(int(row_data['toa']))
            tot.push_back(int(row_data['tot']))
            cal.push_back(int(row_data['cal']))

        # Set the evt value for this entry
        #tree.GetBranch('evt').SetAddress(R.std.vector('int')([int(evt)]))
        
        # Fill the tree with the current event
        tree.Fill()

    # Write the TTree to the ROOT file
    root_file.Write()

    # Close the ROOT file
    root_file.Close()

if __name__=='__main__':
    ## Read dataset
    #df = pd.read_feather(DATA)
    #print(df)

    ## Filter events
    #print("Filtering events...")
    #start = time.time()
    #df = df.groupby('evt').filter(filterEvents)
    #end = time.time()
    #print(f"Elapsed time: {end-start:.3f} s")

    feather_to_root(DATA, f"rootFiles/loop_{args.dataset}.root")
